attestation/src/measurements.rs

```diff
@@ -96,31 +72,80 @@ pub struct ExpectedMeasurements {
+/// across both production and test environments.
+///
+/// The TCB info contains hex-encoded measurement values that are decoded and converted to the
+/// required 48-byte arrays for each measurement register. This provides a single source of truth
+/// for all expected measurements.
+///
+/// TODO(#737): Define a process for updating these static RTMRs going forward, since they are already outdated.
+///
+/// $ git rev-parse HEAD
+/// fbdf2e76fb6bd9142277fdd84809de87d86548ef
+///
+/// See also: https://github.com/Dstack-TEE/meta-dstack?tab=readme-ov-file#reproducible-build-the-guest-image
 impl Default for ExpectedMeasurements {
     fn default() -> Self {
-        // TODO(#737): Define a process for updating these static RTMRs going forward, since they
-        // are already outdated.
-        //
-        // $ git rev-parse HEAD
-        // fbdf2e76fb6bd9142277fdd84809de87d86548ef
-        //
-        // See also: https://github.com/Dstack-TEE/meta-dstack?tab=readme-ov-file#reproducible-build-the-guest-image
-        Self {
+        Self::from_embedded_tcb_info().expect(
+            "Failed to load measurements from embedded TCB info. This indicates a build-time issue with the embedded tcb_info.json file."
+        )
+    }
+}
+
+impl ExpectedMeasurements {
+    /// Attempts to load measurements from the embedded TCB info file.
+    pub fn from_embedded_tcb_info() -> Result<Self, MeasurementsError> {
+        // Parse embedded tcb_info.json file and extract RTMR values dynamically
+        let tcb_info: DstackTcbInfo =
+            serde_json::from_str(TCB_INFO_STRING).map_err(|_| MeasurementsError::InvalidTcbInfo)?;
+
+        // Extract and decode RTMR values from the TCB info (they're in hex format)
+        let measurements = [
+            ("rtmr0", &tcb_info.rtmr0),
+            ("rtmr1", &tcb_info.rtmr1),
+            ("rtmr2", &tcb_info.rtmr2),
+            ("mrtd", &tcb_info.mrtd),
+        ];
+
+        let mut decoded_rtmrs = [[0u8; 48]; 4];
+
+        for (i, (name, hex_value)) in measurements.iter().enumerate() {
+            let decoded = hex::decode(hex_value).map_err(|_| {
+                MeasurementsError::InvalidHexValue(String::from(*name), (*hex_value).clone())
+            })?;
+
+            if decoded.len() != 48 {
+                return Err(MeasurementsError::InvalidLength(
+                    String::from(*name),
+                    decoded.len(),
+                ));
+            }
+
+            decoded_rtmrs[i].copy_from_slice(&decoded);
+        }
+
+        Ok(Self {
             rtmrs: Measurements {
-                mrtd: EXPECTED_MRTD,
-                rtmr0: EXPECTED_RTMR0,
-                rtmr1: EXPECTED_RTMR1,
-                rtmr2: EXPECTED_RTMR2,
+                rtmr0: decoded_rtmrs[0],
+                rtmr1: decoded_rtmrs[1],
+                rtmr2: decoded_rtmrs[2],
+                mrtd: decoded_rtmrs[3],
```

Comment:
Array slicing can panic here as well. Prefer `.get`.

Also we can avoid this with the helper method I suggested.